code/modules/power/smes.dm

```diff
@@ -206,156 +206,83 @@
-	<br><strong>Input level:</strong> <a href='?src=\ref[src];input=set'>[chargelevel]</a> (<a href='?src=\ref[src];input=min'>Min</a> &middot; <a href='?src=\ref[src];input=max'>Max</a>)
-	<div class='bar-outer'><div class='bar'><div class='inner' style="width: [max(0, min(100, round(100 * (lastexcess / SMESMAXCHARGELEVEL), 0.01)))]%;"></div><div class='marker' style="left: [round(100 * (chargelevel / SMESMAXCHARGELEVEL), 0.01)]%;"></div></div>
-	<strong>Available:</strong> [round(lastexcess)] W
-<br>
-<br>
-	<strong>Output:</strong> [online ? "<b>Enabled</b> (<a href='?src=\ref[src];online=1'>Disable</a>)" : "<b>Disabled</b> (<a href='?src=\ref[src];online=1'>Enable</a>)"]
-	<br><strong>Output level:</strong> <a href='?src=\ref[src];output=set'>[output]</a> (<a href='?src=\ref[src];output=min'>Min</a> &middot; <a href='?src=\ref[src];output=max'>Max</a>)
-	<div class='bar'><div class='inner' style="width: [min(100, round(100 * (loaddemand / SMESMAXOUTPUT), 0.01))]%;"></div><div class='marker' style="left: [round(100 * (output / SMESMAXCHARGELEVEL), 0.01)]%;"></div></div>
-	<strong>Current load:</strong> [round(loaddemand)] W
-</div>
-"}
-
-	user.Browse(t, "window=smes;size=400x340")
-	onclose(user, "smes")
-	return
-
-/obj/machinery/power/smes/Topic(href, href_list)
-	..()
-
-	if (usr.stat || usr.restrained() )
+/obj/machinery/power/smes/ui_state(mob/user)
+	return tgui_default_state
+
+/obj/machinery/power/smes/ui_status(mob/user, datum/ui_state/state)
+	return min(
+		state.can_use_topic(src, user),
+		tgui_broken_state.can_use_topic(src, user),
+		tgui_not_incapacitated_state.can_use_topic(src, user)
+	)
+
+/obj/machinery/power/smes/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Smes", src.name)
+		ui.open()
+
+/obj/machinery/power/smes/ui_data(mob/user)
+	var/list/data = list()
+	data["capacityPercent"] = (src.charge / src.capacity * 100)
+	data["charge"] = src.charge
+	data["formattedCharge"] = engineering_notation(src.charge)
+	data["inputAttempt"] = src.chargemode
+	data["inputting"] = src.charging
+	data["inputStartUp"] = src.chargecount //smes waits before charging
+	data["inputLevel"] = src.chargelevel
+	data["inputLevelMax"] = SMESMAXCHARGELEVEL
+	data["inputAvailable"] = src.lastexcess
+	data["outputAttempt"] = src.online
+	data["outputting"] = src.loaddemand
+	data["outputLevel"] = src.output
+	data["outputLevelMax"] = SMESMAXOUTPUT
+	return data
+
+/obj/machinery/power/smes/ui_act(action, params)
+	if(..())
 		return
-
-	if (( usr.using_dialog_of(src) && ((get_dist(src, usr) <= 1) && istype(src.loc, /turf))) || (isAI(usr) || issilicon(usr)))
-		if (href_list["close"])
-			usr.Browse(null, "window=smes")
-			src.remove_dialog(usr)
-			return
-
-		else if ( href_list["cmode"] )
-			chargemode = !chargemode
+	switch(action)
+		if("tryinput") //toggle charging
```

Comment:
`tryinput` is a weird name for this action. Shouldn't this be something like `toggle-input`?